File: scripts/md_to_simple_pdf.mjs
import fs from 'fs';
import path from 'path';

const PAGE_WIDTH = 612;
const PAGE_HEIGHT = 792;
const MARGIN = 72;
const FONT_SIZE = 11;
const LINE_HEIGHT = 14;
const MAX_CHARS = 90;

function wrapLine(line, maxChars) {
  if (line.length <= maxChars) return [line];
  const words = line.split(' ');
  const lines = [];
  let current = '';
  for (const word of words) {
    if (!current) {
      current = word;
    } else if (current.length + 1 + word.length <= maxChars) {
      current += ` ${word}`;
    } else {
      lines.push(current);
      current = word;
    }
  }
  if (current) lines.push(current);
  return lines;
}

function escapePdfText(text) {
  return text.replace(/\\/g, '\\\\').replace(/\(/g, '\\(').replace(/\)/g, '\\)');
}

function buildPdf(lines, outputPath) {
  const usableHeight = PAGE_HEIGHT - 2 * MARGIN;
  const linesPerPage = Math.floor(usableHeight / LINE_HEIGHT);
  const pages = [];
  for (let i = 0; i < lines.length; i += linesPerPage) {
    pages.push(lines.slice(i, i + linesPerPage));
  }

  const objects = [];
  const xrefPositions = [];
  let offset = 0;

  const addObject = (objStr) => {
    xrefPositions.push(offset);
    objects.push(objStr);
    offset += Buffer.byteLength(objStr, 'latin1');
  };

  const header = '%PDF-1.4\n';
  offset += Buffer.byteLength(header, 'latin1');

  addObject('5 0 obj\n<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>\nendobj\n');

  const pageObjects = [];
  const contentObjects = [];
  let objIndex = 6;

  for (const pageLines of pages) {
    const contentStream = ['BT', `/F1 ${FONT_SIZE} Tf`, `${MARGIN} ${PAGE_HEIGHT - MARGIN} Td`];
    for (const line of pageLines) {
      contentStream.push(`(${escapePdfText(line)}) Tj`);
      contentStream.push(`0 ${-LINE_HEIGHT} Td`);
    }
    contentStream.push('ET');
    const contentData = contentStream.join('\n');
    const contentObj = `${objIndex} 0 obj\n<< /Length ${Buffer.byteLength(contentData, 'latin1')} >>\nstream\n${contentData}\nendstream\nendobj\n`;
    contentObjects.push(objIndex);
    addObject(contentObj);
    objIndex += 1;

    const pageObj =
      `${objIndex} 0 obj\n` +
      `<< /Type /Page /Parent 2 0 R /MediaBox [0 0 ${PAGE_WIDTH} ${PAGE_HEIGHT}] ` +
      `/Resources << /Font << /F1 5 0 R >> >> /Contents ${contentObjects[contentObjects.length - 1]} 0 R >>\n` +
      `endobj\n`;
    pageObjects.push(objIndex);
    addObject(pageObj);
    objIndex += 1;
  }

  const kids = pageObjects.map((pid) => `${pid} 0 R`).join(' ');
  addObject(`2 0 obj\n<< /Type /Pages /Count ${pageObjects.length} /Kids [${kids}] >>\nendobj\n`);
  addObject('1 0 obj\n<< /Type /Catalog /Pages 2 0 R >>\nendobj\n');

  const xrefOffset = offset;
  const xref = ['xref', `0 ${objects.length + 1}`, '0000000000 65535 f '];
  for (const pos of xrefPositions) {
    xref.push(`${pos.toString().padStart(10, '0')} 00000 n `);
  }
  const xrefData = `${xref.join('\n')}\n`;

  const trailer =
    `trailer\n<< /Size ${objects.length + 1} /Root 1 0 R >>\n` +
    `startxref\n${xrefOffset}\n%%EOF\n`;

  fs.writeFileSync(outputPath, header, 'latin1');
  for (const obj of objects) {
    fs.appendFileSync(outputPath, obj, 'latin1');
  }
  fs.appendFileSync(outputPath, xrefData, 'latin1');
  fs.appendFileSync(outputPath, trailer, 'latin1');
}

function main() {
  const [, , inputFile, outputFile] = process.argv;
  if (!inputFile || !outputFile) {
    console.error('Usage: md_to_simple_pdf.mjs <input.md> <output.pdf>');
    process.exit(1);
  }
  const raw = fs.readFileSync(path.resolve(inputFile), 'utf8').split(/\r?\n/);
  const lines = [];
  for (const line of raw) {
    if (!line.trim()) {
      lines.push('');
      continue;
    }
    for (const wrapped of wrapLine(line, MAX_CHARS)) {
      lines.push(wrapped);
    }
  }
  buildPdf(lines, path.resolve(outputFile));
  console.log(`Wrote ${outputFile}`);
}

main();

File: scripts/md_to_simple_pdf.py
import sys
from pathlib import Path

PAGE_WIDTH = 612  # Letter
PAGE_HEIGHT = 792
MARGIN = 72
FONT_SIZE = 11
LINE_HEIGHT = 14
MAX_CHARS = 90


def wrap_line(line: str, max_chars: int) -> list[str]:
    if len(line) <= max_chars:
        return [line]
    words = line.split(' ')
    lines = []
    current = ""
    for word in words:
        if not current:
            current = word
        elif len(current) + 1 + len(word) <= max_chars:
            current += " " + word
        else:
            lines.append(current)
            current = word
    if current:
        lines.append(current)
    return lines


def escape_pdf_text(text: str) -> str:
    return text.replace("\\", "\\\\").replace("(", "\\(").replace(")", "\\)")


def build_pdf(lines: list[str], output_path: Path):
    usable_height = PAGE_HEIGHT - 2 * MARGIN
    lines_per_page = int(usable_height // LINE_HEIGHT)
    pages = [lines[i:i + lines_per_page] for i in range(0, len(lines), lines_per_page)]

    objects = []
    xref_positions = []
    offset = 0

    def add_object(obj_str: str):
        nonlocal offset
        xref_positions.append(offset)
        objects.append(obj_str)
        offset += len(obj_str.encode("latin1"))

    # PDF header
    header = "%PDF-1.4\n"
    offset += len(header.encode("latin1"))

    # Font object (5)
    add_object("5 0 obj\n<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>\nendobj\n")

    # Page and content objects
    page_objects = []
    content_objects = []
    obj_index = 6

    for page_lines in pages:
        content_stream = ["BT", f"/F1 {FONT_SIZE} Tf", f"{MARGIN} {PAGE_HEIGHT - MARGIN} Td"]
        for line in page_lines:
            content_stream.append(f"({escape_pdf_text(line)}) Tj")
            content_stream.append(f"0 {-LINE_HEIGHT} Td")
        content_stream.append("ET")
        content_data = "\n".join(content_stream)
        content_obj = f"{obj_index} 0 obj\n<< /Length {len(content_data.encode('latin1'))} >>\nstream\n{content_data}\nendstream\nendobj\n"
        content_objects.append(obj_index)
        add_object(content_obj)
        obj_index += 1

        page_obj = (
            f"{obj_index} 0 obj\n"
            f"<< /Type /Page /Parent 2 0 R /MediaBox [0 0 {PAGE_WIDTH} {PAGE_HEIGHT}] "
            f"/Resources << /Font << /F1 5 0 R >> >> /Contents {content_objects[-1]} 0 R >>\n"
            f"endobj\n"
        )
        page_objects.append(obj_index)
        add_object(page_obj)
        obj_index += 1

    # Pages object (2)
    kids = " ".join(f"{pid} 0 R" for pid in page_objects)
    add_object(f"2 0 obj\n<< /Type /Pages /Count {len(page_objects)} /Kids [{kids}] >>\nendobj\n")

    # Catalog object (1)
    add_object("1 0 obj\n<< /Type /Catalog /Pages 2 0 R >>\nendobj\n")

    # Build xref
    xref_offset = offset
    xref = ["xref", f"0 {len(objects) + 1}", "0000000000 65535 f "]
    for pos in xref_positions:
        xref.append(f"{pos:010d} 00000 n ")
    xref_data = "\n".join(xref) + "\n"

    trailer = (
        "trailer\n"
        f"<< /Size {len(objects) + 1} /Root 1 0 R >>\n"
        "startxref\n"
        f"{xref_offset}\n"
        "%%EOF\n"
    )

    with output_path.open("wb") as f:
        f.write(header.encode("latin1"))
        for obj in objects:
            f.write(obj.encode("latin1"))
        f.write(xref_data.encode("latin1"))
        f.write(trailer.encode("latin1"))


def main():
    if len(sys.argv) != 3:
        print("Usage: md_to_simple_pdf.py <input.md> <output.pdf>")
        sys.exit(1)
    input_path = Path(sys.argv[1])
    output_path = Path(sys.argv[2])
    raw_lines = input_path.read_text(encoding="utf-8").splitlines()
    lines = []
    for line in raw_lines:
        if not line.strip():
            lines.append("")
            continue
        for wrapped in wrap_line(line, MAX_CHARS):
            lines.append(wrapped)
    build_pdf(lines, output_path)
    print(f"Wrote {output_path}")


if __name__ == "__main__":
    main()
